src/test/ui/specialization/issue-39618.rs

```diff
@@ -0,0 +1,24 @@
+// Regression test for #39618, shouldn't crash.
+// check-pass
+
+#![feature(specialization)]
+
+trait Foo {
+    fn foo(&self);
+}
+
+trait Bar {
+    fn bar(&self);
+}
+
+impl<T> Bar for T where T: Foo {
+    fn bar(&self) {}
+}
+
+impl<T> Foo for T where T: Bar {
+    fn foo(&self) {}
+}
+
+impl Foo for u64 {}
```

Comment:
This looks suspicions... There's no `default` anywhere, yet logically, we have `u64: Foo`, hence `where T: Foo` is discharged, hence `u64: Bar`, hence `T: Bar` is discharged, and thus `impl<T> Foo for T` applies (and so we have overlap).

cc @jonas-schievink (I thought this had been fixed?)
cc @matthewjasper & @nikomatsakis